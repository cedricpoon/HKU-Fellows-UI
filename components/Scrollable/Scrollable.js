import React, { PureComponent } from 'react';
import PropTypes from "prop-types";
import * as Animatable from 'react-native-animatable';
import { FlatList as RNFlatList } from 'react-native';

import * as consts from './Constants';

const FlatList = Animatable.createAnimatableComponent(RNFlatList);

class Scrollable extends PureComponent {
  componentWillUnmount() {
    if (this.props.onUnmount) {
      this.props.onUnmount();
    }
  }

  render() {
    const { items, itemRenderer, ...restProps } = this.props

    if (items.length > 0) {
      return (
        <FlatList
          data={items}
          keyExtractor={item => item.id}
          renderItem={itemRenderer}
          animation='fadeIn'
          duration={consts.FADE_IN_DURATION}
          {...restProps}
        />
      );
    } else {
      return null;
    }
  }
}

Scrollable.propTypes = {
  items: PropTypes.array.isRequired,
  itemRenderer: PropTypes.func.isRequired,
  onUnmount: PropTypes.func
};

export default Scrollable;
