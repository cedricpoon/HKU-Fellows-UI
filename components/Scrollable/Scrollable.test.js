import React from 'react';
import { shallow } from 'enzyme';
import Scrollable from './Scrollable';

describe('Testing Scrollable Component', () => {
  const requiredProps = {
    items: [],
    itemRenderer: () => { return null; }
  };

  it('renders as expected with posts', () => {
    const wrapper = shallow(
      <Scrollable {...requiredProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as expected with onUnmount event', () => {
    const wrapper = shallow(
      <Scrollable {...requiredProps} onUnmount={() => {}} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
