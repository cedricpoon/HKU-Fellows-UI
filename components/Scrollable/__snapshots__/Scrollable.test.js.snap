// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Scrollable Component renders as expected with onUnmount event 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Scrollable
    itemRenderer={[Function]}
    items={Array []}
    onUnmount={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Testing Scrollable Component renders as expected with posts 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Scrollable
    itemRenderer={[Function]}
    items={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
